DataStore sync issue from offline saved models (custom type fields)

### Description

This bug report is really close to https://github.com/aws-amplify/amplify-android/issues/1199 from the amplify-android.

I've got a GraphQL model (wich is provided in schema section) that I can save offline or online. When I use the app online, absolutely everything works fine.

When I launch the app online, then turn off the internet connectivity, create and save a new model, it is locally saved properly. Then I end up the program (still offline), turn back the internet and run the program again (online).
The sync is processing, I can retrieve the data from my cloud, but the model that was saved in offline mode is not synced with the cloud.

It actually gives me a mix of DataStore and Amplify exception :

![image](https://user-images.githubusercontent.com/49523387/221905770-9380b018-268b-48e3-8918-9c580855f36a.png)

Regarding the Android issue mentioned above, I think my issue is really close to the same fix but I can't how to overpass it.

The thing that is really curious is, when I run the program, then turn off internet, create and save a model, and turn on internet (without closing the program this time), the model is correctly synced with my cloud.

I use the Auto Conflict resolver.

My Flutter doctor is fine.

If you need more details tell my and I will give you as much as I can.

### Categories

- [ ] Analytics
- [ ] API (REST)
- [X] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Storage

### Steps to Reproduce

1. Use a template wich implement Datastore and AppSync with GraphQL API
2. Create a model that has custom type fields (my custom type has only int and string fields)
3. Run the program online
4. Turn off internet
5. Create and save an object with the model you previously created
6. Close the program (**offline**)
7. Turn the internet back
8. Run the program
9. Here it is

### Screenshots

_No response_

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.3

### Amplify Flutter Version

10.7.3

### Deployment Method

Amplify CLI

### Schema

```GraphQL
# Surgery model is the one where the exception comes from (every custom fields are mentioned in the exception, I've tried to delete them one by one and this is not one field especially).

type Surgery @model @auth(rules: [{allow: owner}]) {
  id: ID!
  localAnesthesia: [LocalAnesthesia]!
  procedure: String
  physicalExamination: PhysicalExamination!
  checkList: CheckList!
  charts: Chart!
}

# The Chart model is an exemple of my custom type field (they all use basics types like int or string)

type Chart {
  timer: Int!
  lowChart: String!
  highChart: String!
}
```

Here's  all I've got:


```
W/amplify:aws-datastore(17941): DataStoreException{message=Error encountered while creating model schema, cause=DataStoreException{message=Failed to get fields for model., cause=AmplifyException{message=An invalid CustomType field was provided. charts must be an instance of SerializedCustomType or a List of instances of SerializedCustomType, cause=null, recoverySuggestion=Check if this model schema is a correct representation of the fields in the provided Object}, recoverySuggestion=Validate your model file.}, recoverySuggestion=See attached exception for more details}
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncClient.create(AppSyncClient.java:147)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor.lambda$create$13$com-amplifyframework-datastore-syncengine-MutationProcessor(MutationProcessor.java:291)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor$$ExternalSyntheticLambda3.publish(Unknown Source:4)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor.lambda$publishWithStrategy$17(MutationProcessor.java:322)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor$$ExternalSyntheticLambda13.subscribe(Unknown Source:4)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:40)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.single.SingleDelayWithObservable$OtherSubscriber.onComplete(SingleDelayWithObservable.java:87)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.single.SingleDelayWithObservable$OtherSubscriber.onNext(SingleDelayWithObservable.java:68)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableTimer$TimerObserver.run(ObservableTimer.java:67)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:41)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:28)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/amplify:aws-datastore(17941): 	at java.lang.Thread.run(Thread.java:923)
W/amplify:aws-datastore(17941): Caused by: DataStoreException{message=Failed to get fields for model., cause=AmplifyException{message=An invalid CustomType field was provided. charts must be an instance of SerializedCustomType or a List of instances of SerializedCustomType, cause=null, recoverySuggestion=Check if this model schema is a correct representation of the fields in the provided Object}, recoverySuggestion=Validate your model file.}
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.buildCreationRequest(AppSyncRequestFactory.java:222)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncClient.create(AppSyncClient.java:144)
W/amplify:aws-datastore(17941): 	... 18 more
W/amplify:aws-datastore(17941): Caused by: AmplifyException{message=An invalid CustomType field was provided. charts must be an instance of SerializedCustomType or a List of instances of SerializedCustomType, cause=null, recoverySuggestion=Check if this model schema is a correct representation of the fields in the provided Object}
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.extractCustomTypeFieldValue(AppSyncRequestFactory.java:547)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.extractFieldValue(AppSyncRequestFactory.java:500)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.extractFieldLevelData(AppSyncRequestFactory.java:427)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.getMapOfFieldNameAndValues(AppSyncRequestFactory.java:392)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.appsync.AppSyncRequestFactory.buildCreationRequest(AppSyncRequestFactory.java:219)
W/amplify:aws-datastore(17941): 	... 19 more

W/amplify:aws-datastore(17941): Error ended observation of mutation outbox:
W/amplify:aws-datastore(17941): DataStoreException{message=Failed to process java.lang.RuntimeException: DataStoreException{message=Error encountered while creating model schema, cause=DataStoreException{message=Failed to get fields for model., cause=AmplifyException{message=An invalid CustomType field was provided. charts must be an instance of SerializedCustomType or a List of instances of SerializedCustomType, cause=null, recoverySuggestion=Check if this model schema is a correct representation of the fields in the provided Object}, recoverySuggestion=Validate your model file.}, recoverySuggestion=See attached exception for more details}, cause=null, recoverySuggestion=Check your internet connection.}
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor.drainMutationOutbox(MutationProcessor.java:122)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor.lambda$startDrainingMutationOutbox$1$com-amplifyframework-datastore-syncengine-MutationProcessor(MutationProcessor.java:103)
W/amplify:aws-datastore(17941): 	at com.amplifyframework.datastore.syncengine.MutationProcessor$$ExternalSyntheticLambda17.apply(Unknown Source:4)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableFlatMapCompletableCompletable$FlatMapCompletableMainObserver.onNext(ObservableFlatMapCompletableCompletable.java:97)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:201)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
W/amplify:aws-datastore(17941): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/amplify:aws-datastore(17941): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/amplify:aws-datastore(17941): 	at java.lang.Thread.run(Thread.java:923)
```

As I can see in the AmplifyDatastore.db that my model is correct (and in fact it is correct because its working with full online).

--- a/packages/amplify/amplify_flutter_android/android/build.gradle
+++ b/packages/amplify/amplify_flutter_android/android/build.gradle
@@ -66,7 +66,7 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation 'com.amplifyframework:core:1.38.2'
+    implementation 'com.amplifyframework:core:1.38.6'
     implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9'
 
     testImplementation 'junit:junit:4.13.2'
@@ -75,6 +75,6 @@ dependencies {
     testImplementation 'androidx.test:core:1.4.0'
     testImplementation 'org.robolectric:robolectric:4.3.1'
     testImplementation 'com.google.code.gson:gson:2.8.6'
-    testImplementation 'com.amplifyframework:aws-auth-cognito:1.38.2'
+    testImplementation 'com.amplifyframework:aws-auth-cognito:1.38.6'
     testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.3.9'
 }

--- a/packages/amplify_core/android/build.gradle
+++ b/packages/amplify_core/android/build.gradle
@@ -63,7 +63,7 @@ android {
 }
 
 dependencies {
-    implementation 'com.amplifyframework:core:1.38.2'
+    implementation 'com.amplifyframework:core:1.38.6'
     implementation 'com.google.code.gson:gson:2.8.6'
     implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9'
 

--- a/packages/amplify_datastore/android/build.gradle
+++ b/packages/amplify_datastore/android/build.gradle
@@ -68,8 +68,8 @@ dependencies {
     // Support for Java 8 features
     coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.1.5'
 
-    implementation "com.amplifyframework:aws-datastore:1.38.2"
-    implementation "com.amplifyframework:aws-api-appsync:1.38.2"
+    implementation "com.amplifyframework:aws-datastore:1.38.6"
+    implementation "com.amplifyframework:aws-api-appsync:1.38.6"
 
     testImplementation 'junit:junit:4.13.2'
     testImplementation 'org.mockito:mockito-core:3.10.0'

--- a/packages/analytics/amplify_analytics_pinpoint_android/android/build.gradle
+++ b/packages/analytics/amplify_analytics_pinpoint_android/android/build.gradle
@@ -66,8 +66,8 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation 'com.amplifyframework:aws-analytics-pinpoint:1.38.2'
-    implementation 'com.amplifyframework:aws-auth-cognito:1.38.2'
+    implementation 'com.amplifyframework:aws-analytics-pinpoint:1.38.6'
+    implementation 'com.amplifyframework:aws-auth-cognito:1.38.6'
 
     testImplementation 'junit:junit:4.13.2'
     testImplementation 'org.mockito:mockito-core:3.10.0'

--- a/packages/api/amplify_api_android/android/build.gradle
+++ b/packages/api/amplify_api_android/android/build.gradle
@@ -66,8 +66,8 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation "com.amplifyframework:aws-api:1.38.2"
-    implementation "com.amplifyframework:aws-api-appsync:1.38.2"
+    implementation "com.amplifyframework:aws-api:1.38.6"
+    implementation "com.amplifyframework:aws-api-appsync:1.38.6"
     implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9'
 
     testImplementation 'junit:junit:4.13.2'

--- a/packages/auth/amplify_auth_cognito_android/android/build.gradle
+++ b/packages/auth/amplify_auth_cognito_android/android/build.gradle
@@ -68,7 +68,7 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation 'com.amplifyframework:aws-auth-cognito:1.38.2'
+    implementation 'com.amplifyframework:aws-auth-cognito:1.38.6'
 
     testImplementation 'junit:junit:4.13.2'
     testImplementation 'org.mockito:mockito-core:3.10.0'

--- a/packages/storage/amplify_storage_s3_android/android/build.gradle
+++ b/packages/storage/amplify_storage_s3_android/android/build.gradle
@@ -56,5 +56,5 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation 'com.amplifyframework:aws-storage-s3:1.38.2'
+    implementation 'com.amplifyframework:aws-storage-s3:1.38.6'
 }

